src/passes/Asyncify.cpp

```diff
@@ -1198,16 +1200,61 @@ struct AsyncifyLocals : public WalkerPass<PostWalker<AsyncifyLocals>> {
   std::unique_ptr<AsyncifyBuilder> builder;
 
   Index rewindIndex;
-  Index numPreservableLocals;
   std::map<Type, Index> fakeCallLocals;
+  std::set<Index> relevantLiveLocals;
+
+  void findRelevantLiveLocals(Function* func) {
+    struct RelevantLiveLocalsWalker
+      : public LivenessWalker<RelevantLiveLocalsWalker,
+                              Visitor<RelevantLiveLocalsWalker>> {
+      // Basic blocks that have a possible unwind/rewind in them.
+      std::set<BasicBlock*> relevantBasicBlocks;
+
+      void visitCall(Call* curr) {
+        if (!currBasicBlock) {
+          return;
+        }
+        // Note blocks where we might unwind/rewind, all of which have a
+        // possible call to ASYNCIFY_CHECK_CALL_INDEX emitted right before the
+        // actual call.
+        // Note that each relevant original call was turned into a sequence of
+        // instructions, one of which is an if and then a call to this special
+        // intrinsic. We rely on the fact that if a local was live at the
+        // original call, it also would be in all that sequence of instructions,
+        // and in particular at the call we look for here (which is right before
+        // the call, and so anything that has its final use at the call is still
+        // live here).
+        if (curr->target == ASYNCIFY_CHECK_CALL_INDEX) {
+          relevantBasicBlocks.insert(currBasicBlock);
+        }
+      }
+    };
+
+    RelevantLiveLocalsWalker walker;
+    walker.walkFunctionInModule(func, getModule());
+    // The relevant live locals are ones that are alive at an unwind/rewind
+    // location. TODO look more precisely inside basic blocks, as one might stop
+    // being live in the middle
```

Comment:
I have a feeling that this TODO is actually pretty important because of how conservative the pass has to be about instrumenting the union of all the locals it finds. How tricky would this be to implement?

Reply:
Not too hard probably, I'll look into it as a followup.

Reply:
This doesn't seem to help, since we add an if around each call - so we've already created a basic block just for the call and nothing else.

Diff for posterity:

```diff
diff --git a/src/passes/Asyncify.cpp b/src/passes/Asyncify.cpp
index d4a788adb..47250cb34 100644
--- a/src/passes/Asyncify.cpp
+++ b/src/passes/Asyncify.cpp
@@ -1226,6 +1226,7 @@ private:
         // live here).
         if (curr->target == ASYNCIFY_CHECK_CALL_INDEX) {
           relevantBasicBlocks.insert(currBasicBlock);
+          currBasicBlock->contents.actions.emplace_back(getCurrentPointer());
         }
       }
     };
@@ -1233,12 +1234,23 @@ private:
     RelevantLiveLocalsWalker walker;
     walker.walkFunctionInModule(func, getModule());
     // The relevant live locals are ones that are alive at an unwind/rewind
-    // location. TODO look more precisely inside basic blocks, as one might stop
-    // being live in the middle
+    // location.
     for (auto* block : walker.liveBlocks) {
       if (walker.relevantBasicBlocks.count(block)) {
-        for (auto local : block->contents.start) {
-          relevantLiveLocals.insert(local);
+        auto liveLocals = block->contents.end;
+        for (auto it = block->contents.actions.rbegin(); it != block->contents.actions.rend(); ++it) {
+          auto& action = *it;
+          if (action.isSet()) {
+            liveLocals.erase(action.index);
+          } else if (action.isGet()) {
+            liveLocals.insert(action.index);
+          } else {
+            // This is a call to check the index, that we noted earlier.
+            assert(action.isOther() && (*action.origin)->is<Call>());
+            for (auto local : liveLocals) {
+              relevantLiveLocals.insert(local);
+            }
+          }
         }
       }
     }
```